Refactor Time Tracking API, Revamp TimeTracking Statistics Page

fm3 opened the pull request:
Changes the time-tracking overview to show times spent in annotations and tasks and filter them by teams and projects. In the linked detail view, the tracked times can also be filtered by type (annotations or tasks) and project.

### Backend Changes
 - Removed several now-unused routes (previously used either by frontend or braintracing dbs)
 - Added new route `/time/overview` to fill new user time tracking table, filtered by team, projectIds, onlyCountTasks
 - Refactoring
   - No more JodaTime dependency, use java.time utilities instead
   - Renaming of some Methods
   - Moved more Logic to SQL
   - Integrated StatisticsController into TimeController

### Steps to test:
- go to https://timetrackingapi.webknossos.xyz/
- Create several users with different rights, annotations, tasks
- Annotate some, as different users
- Time tracking overview should list all users you are allowed to see (only your orga, only yourself if you’re no team manager, only of your teams if you are team manager)
- Use Filters to select specific teams and projects, as well as switch between tasks and non-task annotations.
- Time tracking detail view should list a single user’s spent time in more detail.

### TODOs:
- [x] Backend
  - [x] merge StatisticsController with TimeController?
  - [x] split statistics routes?
  - [x] make inclusion of explorative optional for /api/time routes AND statistics routes (unify those routes altogether?)
  - [x] Move more logic to SQL
  - [x] No more JodaTime
  - [x] Clean up
  - [x] No more java.util.Calendar
  - [x] CI
  - [x] Document changes
  - [x] support selecting multiple teams in `/time/summed/userList`
  - [x] investigate report of empty response depending on team members
  - [x] fix snapshot tests
  - [x] remove unused routes (`api/time/groupedByInterval`?, `/time/allusers`, `/time/userlist`)
  - [x] make sure every user can see their own timetracking info
- [x] Frontend
  - [x] Adapt to changed api for time graph
  - [x] Remove Best/Busiest users of the week table from `webknossos.org/statistics` -> maybe delete whole view
  - [x] Add new timetracking overview table as discussed in https://scm.slack.com/archives/C5AKLAV0B/p1704359796147499
  - [x] adapt timetracking detail view
    - [x] to filter by projects
    - [x] to handle exploratives (task-related entries can now be null)
  - [x] Rethink page structure (time tracking vs statistics)

#### Charlies todos:
**for time tracking overview**
- [x] add and fill table (sortable) in new page
- [x] add CSV export
- [x] link to prefilled detail view
- [x] add filter `include exploratives` 
- [x] test with more than 1 user
- [x] move initial user and daterange values for detail view to url params
- [x] fix key in navbar
- [x] rethink placeholder for projetcs
- [x] time in s in csv
- [x] fix link to detail view, possibly with router
- [x] implement deselect
- [x] include possibility to select only annotations (maybe talk to backend)
- [x] fix DateRangePicker. like in time line view? plain antd? how to do presets without getting TS problems?
- [x] fix spinner
- [x] only visible to admins and team managers
  
**detail view**
- [x] fix project update (somehow an old state is used after getting to view from overview)
- [x] fix google chart, which is not showing time from sth. other than tasks

### Issues:
- fixes #7489
- fixes #5375
- fixes #7707

------
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)

dieknolle3333 commented:
@fm3 would it be possible to have a route that only returns time tracking data for annotations? otherwise i would need to subtract the numbers (tasks and annotations minus tasks) in the frontend.

fm3 commented:
Sure! So we should have something else instead of the onlyCountTasks boolean? Do you have something specific in mind already? A type parameter? How would you name it? Or should we have two booleans and only allow valid combinations?

dieknolle3333 commented:
>So we should have something else instead of the onlyCountTasks boolean? Do you have something specific in mind already? A type parameter? How would you name it? 

Yes, I definitely think we need more than that. I think a type parameter that can have the values "annotations", "tasks" or "annotations&tasks" would be perfect (feel free to adjust the strings if you think they can be abbreviated etc)

fm3 commented:
I see! For this I’d like to reuse this enum https://github.com/scalableminds/webknossos/blob/4f0c3ff28e99807b1f7e3f271e29c74adb495dce/app/models/annotation/AnnotationType.scala (Even though only the two types Task and Explorational make sense here)

So I’d call the parameter `annotationTypes` (plural) and have comma-separated values, e.g. `types=Task`, `types=Task,Explorational`. Would that be ok for you?

dieknolle3333 commented:
@fm3 In the detail view, there is currently a restriction that only time ranges of 31 days or less may be queried. Is this still needed or can I increase the maximum time range?
Other than that, the CI pipeline seems to be failing due to an sbt compile error. Could you maybe have a look there?

fm3 commented:
> 31 days or less

My guess is we can increase the limit, but not indefinitely, to avoid extremely long loads. How about three months? If we find that load times and page performance are not an issue in production (with potentially tens of thousands of timespans), we can increase it further.

> sbt compile error

Maybe you could try to remove the line 6 in `webknossos/app/models/user/time/TimeSpan.scala` (unused import lamented by the CI `import models.annotation.AnnotationType`) to see if that helps. Otherwise I can do it on thursday.

dieknolle3333 commented:
> Maybe you could try to remove the line 6 in webknossos/app/models/user/time/TimeSpan.scala

nice, that was it!

dieknolle3333 commented on .github/common_edge_cases.md:
```diff
@@ -9,7 +9,7 @@ Will this also work if
- Also check that
-  - Complex SQL queries have no fan out effect due to multiple left joins
-  - SQL `IN` statements are never called with empty list
+ - SQL pitfalls:
```
@fm I solved the merge conflicts in this file, feel free to improve them!

dieknolle3333 commented:
@fm I ran into a weird case. I created a new team with only sample user in it. If I send a request with this team id, it returns an empty response. This changes once I add another user to the team. Could there be something wrong or am I missing something and this is supposed to happen?

fm3 commented:
Doesn’t sound right. I’ll have a look tomorrow!

dieknolle3333 commented on conf/application.conf:
```diff
@@ -146,7 +146,7 @@ features {
   discussionBoardRequiresAdmin = false
   hideNavbarLogin = false
-  isWkorgInstance = false
+  isWkorgInstance = true
```
eieiei

dieknolle3333 commented:
I found something else, sorry. If my sample user is in two organizations simultaneously, the route `/api/time/summed/userList` seems to be returning users from the other organization. I made sure that the teams are fetched correctly in the dropdown, and I think they are. So I think that the right team IDs are passed to the request. The user list in the time tracking detail view is also fetched correctly.
Again, sorry if I am missing something.

philippotto commented on frontend/javascripts/admin/statistic/time_tracking_overview.tsx:
```diff
@@ -0,0 +1,261 @@
+      );
+      const allTimeEntries: TimeEntry[] = await Request.receiveJSON(timeEntriesURL);
+      setIsFetching(false);
+      return [allTeams, allProjects, allTimeEntries];
```
is `allTimeEntries` only used for initializing `filteredTimeEntries`? wouldn't this be overwritten directly by the other fetch request? could the initial value be null or [] instead?

philippotto commented on frontend/javascripts/admin/statistic/time_tracking_overview.tsx:
```diff
@@ -0,0 +1,261 @@
+        style={filterStyle}
+        value={[startDate, endDate]}
+        presets={rangePresets}
+        // TODO fix type error. see time_line_view: type error is commented out.
```
todo

dieknolle3333 replied:
Do you have an idea how to do this? The problem is that antd's dayjs and dayjs arent fully matching. Didn't really find a solution online, so I think about just ignoring the type checker there :sweat_smile: 

`Type '(dates: null | (Dayjs | null)[]) => void' is not assignable to type '(dates: NoUndefinedRangeValueType<Dayjs>, dateStrings: [string, string]) => void'.
  Types of parameters 'dates' and 'dates' are incompatible.
    Type 'NoUndefinedRangeValueType<Dayjs>' is not assignable to type '(Dayjs | null)[]'.
      Type 'import("[...]webknossos/node_modules/antd/node_modules/dayjs/index").Dayjs | null' is not assignable to type 'import("[...]/webknossos/node_modules/dayjs/index").Dayjs | null'.
        Type 'Dayjs' is missing the following properties from type 'Dayjs': fromNow, from, toNow, to, and 6 more.ts(2322)`

dieknolle3333 replied:
I managed to fix it by changing the dayjs import to `import dayjs, { Dayjs } from "antd/node_modules/dayjs";`. Not sure whether thats clean though.

philippotto replied:
sounds good to me :)

fm3 commented:
I hope I fixed both your described problems, well spotted! My SQL query was actually missing part of the where clause :sweat_smile: Also, since the sample user is an admin, they are not explicitly assigned to any teams, which is why they were not returned when filtering by teams. I added a line to include orga admins now.

Please try again!

philippotto reviewed (changes requested):
cool stuff! didn't test yet, but already left a couple of comments :)

fm3 commented on conf/webknossos.latest.routes:
```diff
@@ -257,9 +249,12 @@ POST          /termsOfService/accept
-GET           /time/allusers/:year/:month                                                       controllers.TimeController.getWorkingHoursOfAllUsers(year: Int, month: Int, startDay: Option[Int], endDay: Option[Int])
-GET           /time/userlist/:year/:month                                                       controllers.TimeController.getWorkingHoursOfUsers(email: String, year: Int, month: Int, startDay: Option[Int], endDay: Option[Int])
-GET           /time/user/:userId                                                                controllers.TimeController.getWorkingHoursOfUser(userId: String, startDate: Long, endDate: Long)
+GET           /time/allusers/:year/:month                                                       controllers.TimeController.timeSpansOfAllUsers(year: Int, month: Int, startDay: Option[Int], endDay: Option[Int])
+GET           /time/userlist/:year/:month                                                       controllers.TimeController.timeSpansOfUsers(email: String, year: Int, month: Int, startDay: Option[Int], endDay: Option[Int])
```
True, but these routes are still used by external applications afaik, so I don’t want to change them now. Their diff in this PR is only the function naming in the implementation.

philippotto replied:
alright :)

fm3 commented:
@dieknolle3333 am I seeing it correctly that `/time/groupedByInterval` is now unused? (I think this provided the data for the graph that was removed).

If so, I could remove that backend route completely, saving complexity.

philippotto commented:
A question that just came to my mind: non-admin users can still check their own tracked time, right?

fm3 commented:
> non-admin users can still check their own tracked time, right?

Not with the current implementation 🤔  But I guess you are right, we should maybe restore that. @normanrz what do you think?

dieknolle3333 commented:
Well, currently in this PR non-admin users don't see this overview, but they can access the detail view, in which they can see themselves (and not select any other user). Does that suffice? 
![image](https://github.com/scalableminds/webknossos/assets/57364771/4ad06118-6a2a-459c-a54e-ae519d080b8b)

dieknolle3333 commented on frontend/javascripts/admin/statistic/project_and_annotation_type_dropdown.tsx:
```diff
@@ -0,0 +1,110 @@
+  }, [selectedProjectIds, selectedAnnotationType]);
+
+  useEffect(() => {
+    const annotationTypeFilters: NestedSelectOptions = {
+      label: "Filter types",
+      options: [
+        { label: "Tasks & Annotations", value: AnnotationTypeFilterEnum.TASKS_AND_ANNOTATIONS_KEY },
+        { label: "Annotations", value: AnnotationTypeFilterEnum.ONLY_ANNOTATIONS_KEY },
+        { label: "Tasks", value: AnnotationTypeFilterEnum.ONLY_TASKS_KEY },
+      ],
+    };
```
I totally agree. But somehow, it seems to be null initially and wk doesnt render any content (completely white screen). Its probably really obvious why it doesnt work but I don't get it rn

dieknolle3333 replied:
cyclic dependencies!

dieknolle3333 commented on frontend/javascripts/admin/statistic/project_and_annotation_type_dropdown.tsx:
```diff
@@ -0,0 +1,110 @@
+  setSelectedAnnotationType,
+  style,
+}: ProjectAndTypeDropdownProps) {
+  // This state property is an incomplete union of selectedProjectIds and selectedAnnotaionType.
+  // It is mainly used to determine the selected items in the multiselect form item.
+  const [selectedProjectOrTypeFilters, setSelectedProjectOrTypeFilters] = useState(Array<string>);
+  const [taskFilterOptions, setTaskFilterOptions] = useState<Array<NestedSelectOptions>>([]);
```
okay! concerning the comment: Its only the list of project ids that can be empty. I will try to make it more clear anyway :)

dieknolle3333 commented on frontend/javascripts/admin/time/time_line_view.tsx:
```diff
@@ -89,30 +101,94 @@ class TimeLineView extends React.PureComponent<Props, State> {
+    let annotationType = AnnotationTypeFilterEnum.TASKS_AND_ANNOTATIONS_KEY;
+    if (_.has(params, "projectsOrType")) {
+      const projectsOrTypeParam = params.projectsOrType;
+      if (Object.values<string>(AnnotationTypeFilterEnum).includes(projectsOrTypeParam)) {
```
typescript is complaining `Argument of type 'string' is not assignable to parameter of type 'AnnotationTypeFilterEnum'.`

dieknolle3333 commented on frontend/javascripts/admin/time/time_line_view.tsx:
```diff
@@ -89,30 +101,94 @@ class TimeLineView extends React.PureComponent<Props, State> {
+  async componentDidMount() {
+    const isAdminOrTeamManger = isUserAdminOrTeamManager(this.props.activeUser);
+    this.parseQueryParams();
     if (isAdminOrTeamManger) {
-      this.fetchData();
+      await this.fetchData();
+      await this.fetchTimeTrackingData();
```
that seems to work, thank you!

dieknolle3333 commented on frontend/javascripts/admin/time/time_line_view.tsx:
```diff
@@ -89,30 +101,94 @@ class TimeLineView extends React.PureComponent<Props, State> {
     const users = await getEditableUsers();
+    const currentUser =
+      this.state.initialUserId != null && isUserAdminOrTeamManager(this.props.activeUser)
+        ? await getUser(this.state.initialUserId).catch((err) => {
+            Toast.error("User not found.");
+            console.error(err.messages[0].error); // TODO is there a more stable way?
+            return null;
+          })
```
right. I added this because I was running into this case a lot (with this organization-issue that florian has already fixed) and I didnt know what was going on at all. This shouldnt really happen all that often now, so I'll remove it :)

dieknolle3333 commented on frontend/javascripts/admin/statistic/time_tracking_overview.tsx:
```diff
@@ -0,0 +1,261 @@
+  const [selectedTeams, setSelectedTeams] = useState(allTeams.map((team) => team.id));
+  const [projectOrTypeQueryParam, setProjectOrTypeQueryParam] = useState("");
+  useEffect(() => {
+    if (selectedProjectIds.length > 0 && selectedProjectIds.length < allProjects.length) {
```
never, but they can be the same length if all projects are selected. In that case I want to use the `project` type instead of listing all project ids as the query param

philippotto replied:
ahh, I see :+1:

dieknolle3333 commented:
@fm thank you for looking into it! I think that my problems were not completely fixed. The user list doesnt contain any people that arent in the active orga, but not all users are shown, even if they just annotated something. It seems that only the organization owners are shown in the list. Hope that I am not missing sth!

dieknolle3333 commented on CHANGELOG.unreleased.md:
```diff
@@ -33,7 +33,8 @@ For upgrade instructions, please check the [migration guide](MIGRATIONS.released
 - If the current dataset folder in the dashboard cannot be found (e.g., because somebody else deleted it), the page navigates to the root folder automatically. [#7669](https://github.com/scalableminds/webknossos/pull/7669)
-- Voxelytics logs are now stored by organization name, rather than id, in Loki. This is inpreparation of the unificaction of these two concepts. [#7687](https://github.com/scalableminds/webknossos/pull/7687)
+- Voxelytics logs are now stored by organization name, rather than id, in Loki. This is in preparation of the unificaction of these two concepts. [#7687](https://github.com/scalableminds/webknossos/pull/7687)
+- Changed the time-tracking overview to show times spent in annotations and tasks and filter them by teams and projects. In the linked detail view, the tracked times can also be filtered by type (annotations or tasks) and project. [#7524](https://github.com/scalableminds/webknossos/pull/7524)
```
@fm3 feel free to adapt!

philippotto commented on package.json:
```diff
@@ -96,7 +96,8 @@
     "remove-all-snapshots": "rm -rf frontend/javascripts/test/snapshots/",
-    "refresh-e2e-snapshots": "yarn remove-e2e-snapshots && mkdir -p frontend/javascripts/test/snapshots/type-check && docker compose down && docker compose up e2e-tests",
+    "clear-cmake-cache": "rm -f webknossos-jni/target/native/x86_64-linux/build/CMakeCache.txt",
+    "refresh-e2e-snapshots": "yarn remove-e2e-snapshots && mkdir -p frontend/javascripts/test/snapshots/type-check && yarn clear-cmake-cache && docker compose down && docker compose up e2e-tests && yarn clear-cmake-cache",
```
why is the cmake cache clared when refreshing the e2e snapshots? ideally, the docker image wouldn't change `webknossos-jni`. can we "unmount" it?

fm3 replied:
I now see that @frcroth also wrote an issue for this: https://github.com/scalableminds/webknossos/issues/7707

philippotto replied:
ok, makes sense :+1: not sure how much time the cache usually saves, but an alternative would be to rename the cache to `*.bak*` before refreshing the snapshots and afterwards renaming it back (thus, overwriting the cache that was created in between).

fm3 replied:
Good point! I changed it to do just that now :)

philippotto commented on frontend/javascripts/admin/time/time_line_view.tsx:
```diff
@@ -448,6 +431,7 @@ class TimeLineView extends React.PureComponent<Props, State> {
                   }}
                   value={dateRange}
                   // @ts-expect-error ts-migrate(2322) FIXME: Type '(dates: DateRange) => Promise<void>' is not ... Remove this comment to see the full error message
+                  // TODO maybe also apply fix here
```
todo

dieknolle3333 commented on webpack.config.js:
```diff
@@ -198,10 +198,7 @@ module.exports = function (env = {}) {
-          warnings: false,
-          errors: true,
-        },
+        overlay: false,
```
todo: probably remove, find fix or merge intentionally.

fm3 replied:
for the record, we agreed to remove the overlay, so this todo is resolved :)

dieknolle3333 commented:
@fm3 personally I think everything works as expected with your new changes, thank you!

normanrz reviewed (approved):
I reviewed the backend code only. Looks good. Just one comment on time zones. Not sure we have considered these before.

I did a happy path testing. Seems to work. The UI looks slick!

fm3 commented on app/models/user/time/Interval.scala:
```diff
@@ -17,12 +20,12 @@ case class Month(month: Int, year: Int) extends Interval with Ordered[Month] {
+    Instant.fromZonedDateTime(asZonedDateTime.`with`(TemporalAdjusters.lastDayOfMonth()).`with`(LocalTime.MAX))
 
-  private def jodaMonth: DateTime = new DateTime().withYear(year).withMonthOfYear(month)
+  private def asZonedDateTime: ZonedDateTime = ZonedDateTime.of(year, month, 1, 0, 0, 0, 0, ZoneId.systemDefault())
```
It’s the server’s local time, so that shouldn’t be a problem. This is the same behavior as before. I’m not opposed to switching to UTC though, since the server’s timezone is somewhat arbitrary, at least for internationally used instances.

Note that this month-grouping only happens for `/users/:id/loggedTime`, which is not used by the frontend, but only by the python client. Not sure if anybody uses that at all at the moment.

normanrz replied:
I think switching to UTC would be good.

fm3 replied:
switched to UTC now. Still uses the ZonedDateTime class to have the fancy TemporalAdjusters available, but uses ZoneId.of("UTC") now.

philippotto commented on frontend/javascripts/admin/admin_rest_api.ts:
```diff
@@ -1989,6 +1974,21 @@ export async function getTimeTrackingForUser(
+  selectedTypes: AnnotationTypeFilterEnum,
+  projectIds: string[],
+) {
+  // Omit empty parameters in request
+  const projectsParam = projectIds.length > 0 ? `&projectIds=${projectIds.join(",")}` : "";
+  const teamsParam = teamIds.length > 0 ? `&teamIds=${teamIds.join(",")}` : "";
+  return await Request.receiveJSON(
+    `api/time/overview?start=${startMs}&end=${endMs}&annotationTypes=${selectedTypes}${teamsParam}${projectsParam}`,
+  );
```
`URLSearchParams` would be useful here :)

philippotto reviewed (commented):
testing went smooth :) only one small thing:

the bullet points next to the values in the per-user-time-view look weird: 
![image](https://github.com/scalableminds/webknossos/assets/2486553/4960a9c3-0a14-457a-a66b-c5ae2e4d96b4)

maybe use a border-less table and the • symbol in front of each label. or do it like here: https://master.webknossos.xyz/statistics

philippotto reviewed (approved):
front-end looks good to me :1st_place_medal:

fm3 commented on app/controllers/TimeController.scala:
```diff
@@ -1,114 +1,114 @@
 
-    sDate.setTimeInMillis(startDate)
-    eDate.setTimeInMillis(endDate)
+  private def parseCommaSeparated[T](commaSeparatedStrOpt: Option[String])(parseEntry: String => Fox[T]): Fox[List[T]] =
```
No, only parseArrayLiteral in SqlEscaping, but that’s quite a bit more involved, with escaping and quotes and stuff, which is not needed here.

I now moved it from TimeController to TextUtils, though, and the individual ones to ObjectId.fromCommaSeparated and ExtendedEnumeration.fromCommaSeparated.
